corehq/apps/app_manager/models.py

```diff
@@ -6282,34 +6281,44 @@ def app_uses_usercase(app):
             })
         return errors
 
-    def validate_app(self):
-        xmlns_count = defaultdict(int)
-        errors = []
-
-        for lang in self.langs:
-            if not lang:
-                errors.append({'type': 'empty lang'})
-
+    @time_method()
+    def _check_modules(self):
         if not self.modules:
-            errors.append({'type': "no modules"})
+            yield {'type': "no modules"}
         for module in self.get_modules():
             try:
-                errors.extend(module.validate_for_build())
+                for error in module.validate_for_build():
+                    yield error
             except ModuleNotFoundException as ex:
-                errors.append({
+                yield {
                     "type": "missing module",
                     "message": six.text_type(ex)
-                })
+                }
 
+    @time_method()
+    def _check_forms(self):
+        xmlns_count = defaultdict(int)
         for form in self.get_forms():
-            errors.extend(form.validate_for_build(validate_module=False))
+            for error in form.validate_for_build(validate_module=False):
+                yield error
 
             # make sure that there aren't duplicate xmlns's
             if not isinstance(form, ShadowForm):
                 xmlns_count[form.xmlns] += 1
             for xmlns in xmlns_count:
                 if xmlns_count[xmlns] > 1:
-                    errors.append({'type': "duplicate xmlns", "xmlns": xmlns})
+                    yield {'type': "duplicate xmlns", "xmlns": xmlns}
+
+    @time_method()
+    def validate_app(self):
+        errors = []
+
+        for lang in self.langs:
+            if not lang:
+                errors.append({'type': 'empty lang'})
+
+        errors.extend(self._check_modules())
```

Comment:
not sure if yielding inside the function and directly extending after its call preserve the benefits of yield. I think effect would be same as not using yield

Reply:
> I think effect would be same as not using yield

That's the idea, actually.  My intention was not to change the behavior, but to make the code easier to follow.  As it happens, however, the `@time_method` decorator can't evaluate the timing of a generator method, so I'm switching it back to a series of lists anyways.